pyrokinetics/gk_code/gk_output.py

```diff
@@ -14,144 +26,338 @@
+    #: Should be overridden in derived classes.
+    _has_physical_units: ClassVar[Tuple[str, ...]] = ()
+
+    def with_units(self, c: ConventionNormalisation):
+        """
+        Apply units to each quantity in turn and return a new ``Coords``.
+        If units are already applied, renormalises according to the convention supplied.
+        """
+        kwargs = {}
+        for key, val in self.items():
+            if val is None:
+                kwargs[key] = None
+                continue
+            if key in self._has_normalised_units:
+                if hasattr(val, "units"):
+                    kwargs[key] = val.to(c)
+                else:
+                    kwargs[key] = val * self.units(key, c)
+                continue
+            if key in self._has_physical_units:
+                if hasattr(val, "units"):
+                    kwargs[key] = val.to(self.units(key, c))
+                else:
+                    kwargs[key] = val * self.units(key, c)
+                continue
+            # Pass everything else through
+            kwargs[key] = val
+        # Pass through the pseudo-field 'dims'
+        if hasattr(self, "dims"):
+            kwargs["dims"] = self.dims
+        return self.__class__(**kwargs)
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        """
+        Shape of quantities. Raises error if all are None.
+        Should be overridden in ``Coords``, where this function makes no sense.
+        """
+        for _, val in self.items():
+            if val is not None:
+                return np.shape(val)
+        raise ValueError("Fields contains no data")
+
+    def _set_and_check_dims(self, dims) -> None:
+        """Set ``dims``, perform checks on the values assigned."""
+        self.dims = dims
+        for key, val in self.items():
+            if val is not None and np.ndim(val) != len(self.dims):
+                raise ValueError(f"Quantity '{key}' has incorrect number of dims")
+
+
+@dataclasses.dataclass
+class Coords(GKOutputArgs):
+    """Utility dataclass type used to pass coordinates to ``GKOutput``"""
+
+    #: 1D grid of radial wave-numbers used in the simulation
+    #: Units of [rhoref ** -1]
+    kx: ArrayLike
+
+    #: 1D grid of bi-normal wave-numbers used in the simulation
+    #: Units of [rhoref ** -1]
+    ky: ArrayLike
+
+    #: 1D grid of time of the simulation output
+    #: Units of [lref / vref]
+    time: ArrayLike
+
+    #: List of species names in the simulation
+    species: Iterable[str]
+
+    #: 1D grid of theta used in the simulation
+    #: Units of [radians]
+    theta: Optional[ArrayLike] = None
+
+    #: TODO document
+    mode: Optional[ArrayLike] = None
+
+    #: 1D grid of the energy in the simulation
+    #: Units of [dimensionless]
+    energy: Optional[ArrayLike] = None
+
+    #: 1D grid of the pitch angle in the simulation
+    #: Units of [dimensionless]
+    pitch: Optional[ArrayLike] = None
```

Comment:
Nothing actually uses energy/pitch angle coords and likely won't unless we add the ability to read distributions functions in. @d7919 Do you think this is likely to happen so worth including?